File: packages/oxlint-plugin-react-doctor/src/plugin/rules/design/no-layout-transition-inline.ts
import { defineRule } from "../../utils/define-rule.js";
import type { Rule } from "../../utils/rule.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { getInlineStyleExpression } from "./utils/get-inline-style-expression.js";
import { getStylePropertyStringValue } from "./utils/get-style-property-string-value.js";
import { getStylePropertyKey } from "./utils/get-style-property-key.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

export const noLayoutTransitionInline = defineRule<Rule>({
  id: "no-layout-transition-inline",
  title: "Animating layout properties",
  tags: ["test-noise"],
  severity: "warn",
  category: "Performance",
  recommendation:
    "Animate `transform` and `opacity` instead, since they're cheap for the browser. For height, animate `grid-template-rows` from `0fr` to `1fr`.",
  create: (context: RuleContext) => ({
    JSXAttribute(node: EsTreeNodeOfType<"JSXAttribute">) {
      const expression = getInlineStyleExpression(node);
      if (!expression) return;

      for (const property of expression.properties ?? []) {
        const key = getStylePropertyKey(property);
        if (key !== "transition" && key !== "transitionProperty") continue;

        const value = getStylePropertyStringValue(property);
        if (!value) continue;

        const lower = value.toLowerCase();
        if (/\ball\b/.test(lower)) continue;

        const layoutMatch = lower.match(
          /\b(?:(?:max|min)-)?(?:width|height)\b|\bpadding(?:-(?:top|right|bottom|left))?\b|\bmargin(?:-(?:top|right|bottom|left))?\b/,
        );
        if (layoutMatch) {
          context.report({
            node: property,
            message: `Your users see janky, stuttering animation because "${layoutMatch[0]}" relayouts the page every frame, so animate transform & opacity instead.`,
          });
        }
      }
    },
  }),
});

File: native/oxlint/rules/no-layout-transition-inline.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{
    AstNode,
    context::{ContextHost, LintContext},
    rule::Rule,
};

#[derive(Debug, Default, Clone)]
pub struct NoLayoutTransitionInline;

declare_oxc_lint!(
    /// Disallow inline transitions of layout properties.
    NoLayoutTransitionInline,
    react_doctor_native,
    perf,
    version = "0.1.0",
    short_description = "Disallow inline layout transitions.",
);

impl Rule for NoLayoutTransitionInline {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_test_noise_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::JSXAttribute(attribute) = node.kind() else {
            return;
        };
        let Some(style) = get_inline_style_object_expression(attribute) else {
            return;
        };
        if matches!(
            ctx.nodes().parent_kind(node.id()),
            AstKind::JSXOpeningElement(opening_element)
                if is_svg_layout_transition_exempt_element(opening_element)
        ) {
            return;
        }
        for property_name in ["transition", "transitionProperty"] {
            let Some((property, value)) =
                get_effective_static_style_property_string_value(style, property_name)
            else {
                continue;
            };
            let normalized_value = value.to_lowercase();
            let mut value_tokens = normalized_value
                .split(|character| character == ',' || is_js_whitespace(character))
                .filter(|token| !token.is_empty());
            if value_tokens.clone().any(|token| token == "all") {
                continue;
            }
            let Some(layout_property) =
                value_tokens.find(|token| is_layout_transition_property(token))
            else {
                continue;
            };
            ctx.diagnostic(
                OxcDiagnostic::warn(format!(
                    "Your users see janky, stuttering animation because \"{layout_property}\" relayouts the page every frame, so animate transform & opacity instead."
                ))
                .with_label(property.span),
            );
        }
    }
}
